Chewy / Elastic search version update

mjankowski opened the pull request:
As discussed here - https://github.com/mastodon/mastodon/pull/24209#issuecomment-1648720866 - we are limited in our ability to update `omniauth_openid_connect` or anything else which wants to be on Faraday 2+, because chewy gem has dependency on older ES gems which are locked to Faraday 1.x.

Opening this as draft to just get it moving. For now I'm pointing to an experimental branch someone has linked from this issue - https://github.com/toptal/chewy/issues/847 - which has started on some basic class changes in Chewy to support changes made in the ES gems. With that alone and one small change on our side I was able to get things loading/running/passing, but there are a ton of caveats here (hence the draft status!):

- Similar to the omniauth changes, we don't have exhaustive chewy/ES integration coverage in the specs. Most of what we do have is just stubbing out ES responses with chewy mock helpers and verifying basic chewy behavior. Makes it hard to be confident about results here. We should almost definitely improve this situation before upgrading.
- How fully does this experimental chewy branch actually do all updates needed for the 8.x ES gems (split into transport and api gems) -- I don't know currently.
- There’s a warning from the gem during spec run because of error connecting to server and how server is not official supported version. I think this is actually fine to ignore, but in the few mins I looked it didn't seem to be disableable.
- ~~There's an insistence from the official 8.x server on secure/443/https connections, which our default config does not seem to do.~~ (not relevant: we wont update server to 8.x any time soon)
- ~~I suspect we have some specs which are not correctly stubbed out … I was able to get different spec output when running -vs- not running an actual ES server locally~~ (fixed in https://github.com/mastodon/mastodon/pull/26413)
- I did a quick console session to see if Chewy could just connect with both a 7.x and 8.x server running locally and this chewy gem update in place, and was able to at least get that connection established, but have done zero work yet to verify anything else works.
- ~~I know there are licensing changes on the server side of things either in server v7 or moving from 7 to 8 and have not at all contemplated whether that matters here yet (I think it does not narrowly matter, because we're not changing the packaged Docker ES distribution, but maybe its related).~~ (not relevant: we arent packaging server, and we will only update client libs here)

All that to say, this is currently focused merely on accomplishing the upgrade at all, and needs a lot more verification of behavior (which I’ll continue on) … just wanted to open this to clarify direction here and whether this is a good path to work on (as opposed to investigating a switch to OpenSearch, and/or any other ideas to open up our Faraday update Future).

mjankowski committed:
Use Chewy PR branch that supports ES 8.x

mjankowski committed:
Rename some ES gem classes

renchap commented:
> * I know there are licensing changes on the server side of things either in server v7 or moving from 7 to 8 and have not at all contemplated whether that matters here yet (I think it does not narrowly matter, because we're not changing the packaged Docker ES distribution, but maybe its related).

IIRC Elastic support versions N and N-1 in their official libs, so we should be able to use their version 8 lib with a v7 server.

The licence change is significant, but in practice it only impacts you if you embed ES, or offer it as a service, which should not impact any Mastodon admin or user.

mjankowski commented:
> IIRC Elastic support versions N and N-1 in their official libs, so we should be able to use their version 8 lib with a v7 server.
> 
> The licence change is significant, but in practice it only impacts you if you embed ES, or offer it as a service, which should not impact any Mastodon admin or user.

Thanks, good to know. I'll keep this PR focused on getting the 8.x client libs from the gems working against a running 7.x ES server cluster.

mjankowski commented:
Rebased against main after the spec stub thing was merged in. Edited original description to strike out some not relevant concerns.

I've done a little more local console testing, and these things seem ok:
- Enable Chewy/ES with env vars; create a status with some unique text value; run `StatusIndex.query(...)`, confirm status is in results
- Run `bin/tootctl search deploy`, output looks fine, ES server logs seem normal during command
- Use web UI to perform similar experiment -- server running in ES mode is able to search/retrieve relevant accounts/statuses from dev db after being indexed.

Would love ideas here for what else to verify.

mjankowski commented:
Converted from Draft to ready for review because I think this might actually be fine.

Would still love ideas for more things to test out (either via added specs or just local console work).

Note that this PR diff currently includes a gem bump to a specicic commit ref on a PR branch, not an official chewy release. Tried to [ping chewy maintainers](https://github.com/toptal/chewy/issues/866#issuecomment-1671295658) to see if we can help get it updated.

mjankowski commented:
There is now a Chewy PR from the maintainers which starts on what I assume will be an approved update here. Rebased this branch, and updated the Gemfile to point to that PR branch instead of the previous one. The actual code changes are minimal for us, and it looks like they are going to do a hard bump where the released version requires the higher ES version.

mjankowski commented:
Rebased and updated elasticsearch again. The Chewy PR branch hasn't made much progress recently, but still moving.

mjankowski commented:
Possible interim path here -- looks like the ES gem in its newer 7.x releases, supports either faraday 1 or 2. If we can get Chewy (still in 7.x series) to relax it's requirement to allow newer ES7 gem, we could get a relaxed faraday req prior to ES8 update on the chewy side. Opened a PR to hopefully get discussion/approval of the idea: https://github.com/toptal/chewy/pull/933

mjankowski committed:
Update elasticsearch to version 8.13.0

mjankowski commented:
Rebased after https://github.com/mastodon/mastodon/pull/30157 merged.

This is still pointing to the WIP PR branch for chewy 8. I'll keep monitoring that and update if/when that releases.

Fortunately, we got the faraday range flexibility from the 7.x branch, so this is less pressing as a result.

mjankowski committed:
Update monkey patch to match signature

mjankowski committed:
Use a different PR branch

mjankowski commented:
The PR branch this had been tracking was closed in favor of another (still WIP) branch. Updated to point to that one. No idea how close/far away it is from being ready.

mjankowski committed:
Update chewy to revision 545118fa11f0d18ef2bad06d88c70ca5b715fcfa

mjankowski committed:
Use `toptal/chewy` revision 84c0eed871e835f55d560097ef141cd5fe504967

mjankowski commented:
The PR on chewy merged on their side, updated this PR to point to `toptal/chewy` main branch.

I assume they'll release an 8.0.0 beta gem at some point, will update again if so.

Related - it looks like the actual ES server (not the client gem) 8.x series is now dual licensed, might make upgrade path more viable for us on that side of things if desired.

mjankowski committed:
Update chewy to version 8.0.0.pre.beta

mjankowski commented:
There's an `8.0.0.pre.beta` release out, using that here. Will attempt to reproduce the CI issues locally.

mjankowski commented:
Possible option here - research alternate adapter, ie - https://github.com/ankane/searchkick

mjankowski commented:
This PR is almost two years old and there is still not Chewy support for ES 8.x client gems. In the meantime the official ES client/api gems have done a whole march through the 8.x series and are into 9.x now. There is a chewy beta gem release from ~9 months ago, but little regular progress since, and many open (not responded to) issues asking about support for newer Rails/ES/etc versions.

Nothing is broken here and so this it not urgent, just wanted to bring it up. I closed this PR because I got tired of rebasing and there did not really seem to an imminent release or anything. In the event they ever release with the support we need, I'll rebase/reopen.

All that said ... I'm curious about any interest in exploring a switch to the `searchkick` gem? https://github.com/ankane/searchkick - these are somewhat vanity metrics, but it has many more forks/stars/etc than chewy, and seems to be much more actively maintained. I'm willing to do a little research/spike branch on this front, but also don't want to waste my time. If it's a non-starter and we're committed to chewy for whatever reason, I won't bother. If we're at least open to it and have time to review/react to an experiment, I'll give it a go.

I've used it successfully on small hobby personal project, but have not really contemplated how/whether/if it could successfully fit into Mastodon or not yet.

renchap commented:
Yes, I would be inclined to move away from Chewy, but I am afraid of the migration. Ideally we would reuse the same ES indices.

But if we do this, I would really really like us to better support schema updates. At the moment schema updates will drop the existing indices, then recreate them with the new schema, and reindex all content from the database.

We should support a mode where new indices are created, content is reindexed (ideally using the ES reindex API so we dont need to read all rows from the DB again), then the new indices start to be used and the old ones are dropped. This is causing real production pain right now as a reindex can take 2 weeks on large servers, and causes a lot of CPU usage
